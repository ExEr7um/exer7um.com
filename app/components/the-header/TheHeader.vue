<script setup lang="ts">
/** Список навигации */
const navigation = [
  "index",
  "experience",
  "projects",
  "about",
] as const satisfies string[]
</script>

<template>
  <header class="container sticky top-4">
    <div
      class="card flex h-20 items-center justify-between gap-x-2 bg-opacity-30 px-4 shadow backdrop-blur-lg md:pl-8"
    >
      <TheHeaderNavigation class="hidden md:flex" :navigation />
      <TheHeaderNavigationMobile class="md:hidden" :navigation />
      <div class="flex grow justify-between gap-x-2 sm:grow-0">
        <TheHeaderLocaleSwitch />
        <UiContactMeButton />
      </div>
    </div>
  </header>
</template>
