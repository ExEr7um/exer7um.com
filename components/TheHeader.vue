<script setup lang="ts"></script>

<template>
  <header></header>
</template>
