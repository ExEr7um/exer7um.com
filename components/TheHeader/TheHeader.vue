<script setup lang="ts">
import {
  Bars3BottomLeftIcon,
  XMarkIcon,
} from "@heroicons/vue/24/outline/index.js"
const localePath = useLocalePath()
const { t } = useI18n()

/** Открыто ли мобильное меню */
const isMenuOpened = ref(false)

/**
 * Функция для смены состояния мобильного меню.
 *
 * При использовании функции, меняется состояние `isMenuOpened` на противоположное.
 */
function toggleMenu() {
  isMenuOpened.value = !isMenuOpened.value
}
</script>

<template>
  <header class="mb-24">
    <div
      class="fixed top-0 z-40 h-24 w-full bg-white transition dark:bg-neutral-800"
      :class="{ 'bg-opacity-0 dark:bg-opacity-0': isMenuOpened }"
    >
      <div
        class="container relative z-40 flex h-full items-center justify-between bg-white dark:bg-neutral-800 sm:transition-all"
        :class="{ 'sm:rounded-t-xl': isMenuOpened }"
      >
        <button class="button-secondary lg:hidden" @click="toggleMenu">
          <XMarkIcon v-if="isMenuOpened" class="h-6 w-6" />
          <Bars3BottomLeftIcon v-else class="h-6 w-6" />
          <span>{{ t("header.menu") }}</span>
        </button>
        <Navigation class="hidden lg:flex" />
        <div class="flex items-center">
          <LocaleSwitch class="hidden md:flex" />
          <NuxtLink :to="localePath('/message')">
            <button class="button-primary" @click="isMenuOpened = false">
              {{ t("header.contactMe") }}
            </button>
          </NuxtLink>
        </div>
      </div>
      <Transition name="slide-down">
        <TheHeaderMobileMenu v-if="isMenuOpened" class="lg:hidden" />
      </Transition>
    </div>
    <Transition>
      <div
        v-if="isMenuOpened"
        class="fixed inset-0 z-30 bg-neutral-900/10 backdrop-blur lg:hidden"
      />
    </Transition>
  </header>
</template>
