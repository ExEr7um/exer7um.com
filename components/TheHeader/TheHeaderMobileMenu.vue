<script setup lang="ts">
const { t } = useI18n()
</script>

<template>
  <div
    class="container relative z-30 flex flex-col gap-y-8 rounded-b-xl bg-white pt-2 pb-10 dark:bg-neutral-800"
  >
    <div class="flex flex-col gap-y-6">
      <h3>{{ t("header.mobileMenu.navigation") }}</h3>
      <Navigation :is-vertical="true" />
    </div>
    <div class="flex flex-col gap-y-4">
      <h3>{{ t("header.mobileMenu.changeLanguage") }}</h3>
      <LocaleSwitch class="-ml-2" />
    </div>
  </div>
</template>
