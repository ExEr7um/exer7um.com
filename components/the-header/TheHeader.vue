<script setup lang="ts">
/** Список навигации */
const navigation = [
  "index",
  "projects",
  "experience",
  "about",
] as const satisfies string[]
</script>

<template>
  <header class="container sticky top-4">
    <div
      class="card flex h-20 items-center justify-between bg-opacity-30 px-4 shadow backdrop-blur-lg md:pl-8"
    >
      <TheHeaderNavigation class="hidden md:flex" :navigation="navigation" />
      <TheHeaderNavigationMobile class="md:hidden" :navigation="navigation" />
      <ContactMeButton />
    </div>
  </header>
</template>
