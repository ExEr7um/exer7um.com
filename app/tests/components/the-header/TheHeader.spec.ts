import type { VueWrapper } from "@vue/test-utils"

import { mountSuspended } from "@nuxt/test-utils/runtime"
import { afterEach, beforeEach, describe, expect, test } from "vitest"

import TheHeader from "~/components/the-header/TheHeader.vue"

describe("Компонент TheHeader", () => {
  let wrapper: VueWrapper

  const desktopNavigation = () =>
    wrapper.findComponent({ name: "UNavigationMenu" })
  const mobileNavigation = () =>
    wrapper.findComponent({ name: "TheHeaderNavigationMobile" })

  beforeEach(async () => {
    wrapper = await mountSuspended(TheHeader, { shallow: true })
  })

  afterEach(() => {
    wrapper.unmount()
  })

  test("Семантический тег header", () => {
    expect(wrapper.element.tagName).toBe("HEADER")
  })

  describe("Десктопная навигация", () => {
    test("Отображается", () => {
      expect(desktopNavigation().exists()).toBeTruthy()
    })

    test("Передаются элементы навигации", () => {
      expect(desktopNavigation().props("items")).toMatchSnapshot()
    })

    test("Отображается на десктопных устройствах", () => {
      expect(desktopNavigation().classes()).toContain("md:flex")
    })
  })

  describe("Мобильная навигация", () => {
    test("Отображается", () => {
      expect(mobileNavigation().exists()).toBeTruthy()
    })

    test("Передаются элементы навигации", () => {
      expect(mobileNavigation().props("navigation")).toMatchSnapshot()
    })

    test("Скрывается на десктопных устройствах", () => {
      expect(mobileNavigation().classes()).toContain("md:hidden")
    })
  })
})
