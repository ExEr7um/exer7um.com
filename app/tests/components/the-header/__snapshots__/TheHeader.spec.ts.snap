// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Компонент TheHeader > Десктопная навигация > Передаются элементы навигации 1`] = `
[
  {
    "icon": "heroicons:home",
    "label": "pages.index",
    "to": "/",
  },
  {
    "icon": "heroicons:briefcase",
    "label": "pages.experience",
    "to": "/experience/",
  },
  {
    "icon": "heroicons:squares-2x2",
    "label": "pages.projects",
    "to": "/projects/",
  },
  {
    "icon": "heroicons:user",
    "label": "pages.about",
    "to": "/about/",
  },
]
`;

exports[`Компонент TheHeader > Мобильная навигация > Передаются элементы навигации 1`] = `
[
  {
    "icon": "heroicons:home",
    "label": "pages.index",
    "to": "/",
  },
  {
    "icon": "heroicons:briefcase",
    "label": "pages.experience",
    "to": "/experience/",
  },
  {
    "icon": "heroicons:squares-2x2",
    "label": "pages.projects",
    "to": "/projects/",
  },
  {
    "icon": "heroicons:user",
    "label": "pages.about",
    "to": "/about/",
  },
]
`;
